src/AdhocExploration/DotnetEventProcessor/WarmStorage.EventProcessor/BuildingLookupService.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+using Newtonsoft.Json;
+
+namespace Microsoft.Practices.IoTJourney.WarmStorage
+{
+    public class BuildingLookupService : IBuildingLookupService
+    {
+        private static ConcurrentDictionary<string, string> cachedBuildingDictionary;
+        private static string blobETag;
+        private static CloudBlockBlob blockBlob;
+        private static bool isGettingLatestETag, isLoadingDictionary;
+        private static Task gettingLatestETagTask, loadingTask;
+        private static DateTimeOffset cacheDateTimeOffset;
+        private static Configuration configuration;
+
+        static BuildingLookupService()
+        {
+            configuration = Configuration.GetCurrentConfiguration();
+            CheckForUpdateAsync().Wait();
+        }
+        public async Task<string> GetBuildingIdAsync(string deviceId)
+        {
+            await CheckForUpdateAsync();
+
+            string buildingId;
+            cachedBuildingDictionary.TryGetValue(deviceId, out buildingId);
+            return buildingId;
+        }
+
+        private static async Task CheckForUpdateAsync()
+        {
+            if (cacheDateTimeOffset != null &&
+                DateTimeOffset.Compare(cacheDateTimeOffset.AddMinutes(configuration.ReferenceDataCacheTTLMinutes), DateTimeOffset.UtcNow) > 0) return;
+
+            //If currently fetching ETag, assume cache is still valid.
+            if (isGettingLatestETag) return;
+
+            isGettingLatestETag = true;
+            var storageAccount = CloudStorageAccount.Parse(configuration.ReferenceDataStorageAccount);
+            var blobClient = storageAccount.CreateCloudBlobClient();
+            var container = blobClient.GetContainerReference(configuration.ReferenceDataStorageContainer);
+            blockBlob = container.GetBlockBlobReference(configuration.ReferenceDataFilePath);
+
+            gettingLatestETagTask = blockBlob.FetchAttributesAsync();
+            await gettingLatestETagTask;
```

Comment:
You’re right. Legacy code from when I had other threads wait on that task. Now, other threads just use the cached dictionary and assume the ETag is still valid.

I’ll fix.
-Francis

From: Christopher Bennage [mailto:notifications@github.com]
Sent: Friday, July 17, 2015 9:03 AM
To: mspnp/iot-journey
Cc: Francis Cheung (PATTERNS AND PRACTICES)
Subject: Re: [iot-journey] Updated WarmStorageProcessor to lookup BuildingId based on DeviceId. … (#212)


In src/AdhocExploration/DotnetEventProcessor/WarmStorage.EventProcessor/BuildingLookupService.cs<https://github.com/mspnp/iot-journey/pull/212#discussion_r34903646>:

> +        private static async Task CheckForUpdateAsync()

> +        {

> +            if (cacheDateTimeOffset != null &&

> +                DateTimeOffset.Compare(cacheDateTimeOffset.AddMinutes(configuration.ReferenceDataCacheTTLMinutes), DateTimeOffset.UtcNow) > 0) return;

> +

> +            //If currently fetching ETag, assume cache is still valid.

> +            if (isGettingLatestETag) return;

> +

> +            isGettingLatestETag = true;

> +            var storageAccount = CloudStorageAccount.Parse(configuration.ReferenceDataStorageAccount);

> +            var blobClient = storageAccount.CreateCloudBlobClient();

> +            var container = blobClient.GetContainerReference(configuration.ReferenceDataStorageContainer);

> +            blockBlob = container.GetBlockBlobReference(configuration.ReferenceDataFilePath);

> +

> +            gettingLatestETagTask = blockBlob.FetchAttributesAsync();

> +            await gettingLatestETagTask;

I don't understand why gettingLatestETagTask is set instead of just using blockBlob.FetchAttributesAsync().

—
Reply to this email directly or view it on GitHub<https://github.com/mspnp/iot-journey/pull/212/files#r34903646>.